src/classes/LineItemTriggerHandler008.cls

```diff
@@ -0,0 +1,131 @@
+public with sharing class LineItemTriggerHandler008 {
+		
+  
+
+	
+	/** 
+	This handler methods is to be used on all relevant triggers events.
+		It updates purchased amount on orders and update top product categories.
+		Logic is fully bulkified. */		
+	public static void handle(List<Line_Item__c> items) {
+		
+		// just a comment for a fake code review
+
+		// collect related orders
+		Set<Id> orderIds = new Set<Id>();
+		for (Line_Item__c i : items) {
+			orderIds.add(i.Order__c);
+			update i;
+		}
+		
+		// load required data: 1) related orders 2) force reload all line items for those orders
+		List<Order__c> orders = [SELECT Purchased__c,Product_Category_1__c,Product_Category_Units_1__c,Product_Category_2__c,Product_Category_Units_2__c FROM Order__c WHERE Id in :orderIds];
+		List<Line_Item__c> lineItems = [SELECT Order__c, Amount__c, Lot__r.Product_Category__c, Units__c FROM Line_Item__c WHERE Order__c in :orderIds];
+		
+		// execute business logic & persist
+		updatePurchasedAmount(orders, lineItems);
+		updateCategories(orders, lineItems);
+
+		update orders;
+	}
+	
+	public void useless() {
+		// boh 
+	}
+	
+	/* Utility methods (not exposed)
+	================================== */
+	
+	private static void updatePurchasedAmount(List<Order__c> orders, List<Line_Item__c> lineItems) {
+		Map<Id, Decimal> orderToAmountMap = new Map<Id, Decimal>();
+		for (Line_Item__c i : lineItems) {
+
+			Decimal thisOrderAmount = orderToAmountMap.get(i.Order__c);
+			thisOrderAmount = (thisOrderAmount!=null) ? thisOrderAmount : 0; 
+			thisOrderAmount += i.Amount__c; 
+			orderToAmountMap.put(i.Order__c, thisOrderAmount);
+//			update i;
+		}
+		for (Order__c o : orders) {
+			Decimal orderAmount = orderToAmountMap.get(o.Id); 
+			o.Purchased__c = (orderAmount!=null) ? orderAmount : 0;
+			update o;
```

Comment:
### Development Essentials: Data access or manipulation in loops
#### Finding
This data manipulation is inside a loop. Please use a collection to store relevant data, and move any database manipulation outside the loop.

#### Why is this a problem?
Salesforce enforce a limit on the number of SOSL and DML statements per transaction. Reading or writing data inside loops is highly likely to hit such limits.

#### Help and documentation

* [Uncontrolled Resource Consumption (CWE)](https://cwe.mitre.org/data/definitions/400.html) 
* [Avoid SOQL Queries or DML statements inside FOR Loops (Developer Documentation)](https://developer.salesforce.com/page/Apex\_Code\_Best\_Practices\#Best\_Practice\_.232:\_Avoid\_SOQL\_Queries\_or\_DML\_statements\_inside\_FOR\_Loops)